### PR TITLE
Update flat_correct to allow passing in a normalization.

### LINKED ISSUE #577: flat_correct: allow passing in a normalization factor
flat_correct automatically calculates the mean of the flat field and normalizes the flat before dividing by it. This is convenient, but if you want to use two different flats on different sets of images but force them to have the same normalization, there is no way to turn off the normalization or pass in an externally determined factor.

This becomes an issue when processing multi-extension FITS files. For example, if extensions 1 and 2 are two different CCDs, or two halves of the same CCD, and one tries to process them by running flat_correct on the two extensions separately, then the flats need to have the same normalization. (Imagine that the left half of the CCD is extension 1 and it is 10% more sensitive than ext 2. Then both the science targets and the flat field will have 10% more counts in ext 1, and the target should be divided down by the brighter flat field. Renormalizing ext 1 and ext 2 separately by their own means will produce a non-flat science image.)

For an example of how I had to work around this behavior, you can see the function _renormalize_by_flat()_ in the pipeline [here](https://github.com/bjweiner/ARTN/blob/master/mont4k_pipeline/kuiper_reduce.py#L509). It would be simpler just to have a normalizing factor passed in as an optional argument to ccdproc.flat_correct. I can write this code and issue a pull request, but someone with more practice will need to read it for style and write a test.

This is a small part of making ccdproc friendly to multi-extension FITS files, but can stand on its own.

### PR DESCRIPTION
**Pull request by bjweiner**
- [x] For new contributors: Did you add yourself to the "Authors.rst" file?
- [x] Did you add an entry to the "Changes.rst" file?
- [x] Did you include a meaningful docstring with Parameters, Returns and Examples?
- [x] Does the commit message include a "Fixes #issue_number" (replace "issue_number"). 
- [x] Did you include tests for the new functionality?
- [x] Does this PR add, rename, move or remove any existing functions or parameters?

Fixes #577

### CODE REVIEW

**Comment by MSeifert04** [comment]
Thank you for making the pull request 👍 

I hope I have some time this weekend to look at it in more detail but regarding the test failure: There are some whitespaces in the "empty lines". Also I would prefer an Exception (ValueError) in case it's <= 0, silently ignoring invalid arguments.

**Comment by mwcraig** [comment]
@bjweiner -- we are hoping to get a release of version 1.3 done otday or tomorrow and I really want to get this change in so I'm adding a test and addressing @MSeifert04's comments. I'll push those to your patch branch in ~15 minutes (and add you to the AUTHORS list). No need for further action on your part.

Normally we'd walk you through revising the PR but we're already ~3 weeks behind on the 1.3 release...

Thanks again for doing this!

**Comment by mwcraig** [comment]
@MSeifert04 -- could you please give this a quick look?

File: ccdproc/core.py [hunk 1]
```diff
@@ -736,6 +737,13 @@ def flat_correct(ccd, flat, min_value=None):
         will replace all values in the flat by the min_value.
         Default is ``None``.
 
+    mean_value : float or None, optional
+        Normalize flat field by this argument rather than the mean of the
```
Comment (MSeifert04) [inline ccdproc/core.py]: Probably should start with an `If not None, normalize flat ....`

File: ccdproc/tests/test_ccdproc.py [hunk 2]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+                               flat.data / flat_mean)
+
+    # check that metadata is unchanged (since logging is turned off)
+    assert flat_data.header == ccd_data.header
+
+
+def test_flat_correct_mean_value_bad_value(ccd_data):
+    # Test that flat_correct raises the appropriate error if
+    # it is given a bad mean_value. Bad means <=0.
+    ccd_data.header['my_key'] = 42
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: The key-value pair is not really essential for this test, right?

File: ccdproc/tests/test_ccdproc.py [hunk 3]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+def test_flat_correct_mean_value(ccd_data):
+    # add metadata to header for a test below...
+    ccd_data.header['my_key'] = 42
+    size = ccd_data.shape[0]
+    # create the flat, with some scatter
+    # Note that mean value of flat is set below and is different than
+    # the mean of the flat data.
+    flat_mean = 5.0
+    data = np.random.normal(loc=1.0, scale=0.05, size=(size, size))
+    flat = CCDData(data, meta=fits.header.Header(), unit=ccd_data.unit)
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: you can omit the `.header`, just using `fits.Header` should work without problems.

File: ccdproc/tests/test_ccdproc.py [hunk 4]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+def test_flat_correct_mean_value_bad_value(ccd_data):
+    # Test that flat_correct raises the appropriate error if
+    # it is given a bad mean_value. Bad means <=0.
+    ccd_data.header['my_key'] = 42
+    size = ccd_data.shape[0]
+    # create the flat, with some scatter
+    data = np.random.normal(loc=1.0, scale=0.05, size=(size, size))
+    flat = CCDData(data, meta=fits.header.Header(), unit=ccd_data.unit)
+    with pytest.raises(ValueError) as e:
+        _ = flat_correct(ccd_data, flat, add_keyword=None,
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: The `_ =` can be omitted.

File: ccdproc/tests/test_ccdproc.py [hunk 5]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+
+
+def test_flat_correct_mean_value_bad_value(ccd_data):
+    # Test that flat_correct raises the appropriate error if
+    # it is given a bad mean_value. Bad means <=0.
+    ccd_data.header['my_key'] = 42
+    size = ccd_data.shape[0]
+    # create the flat, with some scatter
+    data = np.random.normal(loc=1.0, scale=0.05, size=(size, size))
+    flat = CCDData(data, meta=fits.header.Header(), unit=ccd_data.unit)
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: Same here, it doesn't need the `.header`

File: ccdproc/tests/test_ccdproc.py [hunk 6]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+    # create the flat, with some scatter
+    # Note that mean value of flat is set below and is different than
+    # the mean of the flat data.
+    flat_mean = 5.0
+    data = np.random.normal(loc=1.0, scale=0.05, size=(size, size))
+    flat = CCDData(data, meta=fits.header.Header(), unit=ccd_data.unit)
+    flat_data = flat_correct(ccd_data, flat, add_keyword=None,
+                             mean_value=flat_mean)
+
+    #check that the flat was normalized
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: Should have a whitespace between `#` and `check ...`

File: ccdproc/tests/test_ccdproc.py [hunk 7]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
...
+@pytest.mark.data_scale(10)
+def test_flat_correct_mean_value(ccd_data):
+    # add metadata to header for a test below...
+    ccd_data.header['my_key'] = 42
+    size = ccd_data.shape[0]
+    # create the flat, with some scatter
+    # Note that mean value of flat is set below and is different than
+    # the mean of the flat data.
+    flat_mean = 5.0
+    data = np.random.normal(loc=1.0, scale=0.05, size=(size, size))
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: Probably more robust if we use `size=ccd_data.shape` here (and in the other new test).

File: ccdproc/tests/test_ccdproc.py [hunk 8]
```diff
@@ -481,6 +481,47 @@ def test_flat_correct_min_value(ccd_data):
     assert flat_orig_data is not flat.data
 
 
+# test for flat correction
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: I guess this was copied from the test above. I don't think this accurately describes the test case, it probably needs a `with mean_value` appended or we could remove it here altogether.

**Comment by MSeifert04** [comment]
I added some "nitpick-comments", feel free to ignore them :) A lot of these are also present in other tests, so that's no merge-blocker.

However I've been thinking about the argument name and `mean_value` seems a bit too specific. Wouldn't `norm_value` describe the purpose more accurately?

Also if you have the time a changelog entry would be good to have. But we can also add that when we do the next release. :)

**Comment by mwcraig** [comment]
Addressed the review comments and rebased to resolve the changelog conflict.

@bjweiner -- I've changed the name of the argument from `mean_value` to `norm_value` but if you feel strongly that it should be changed back please let us know. We are hoping to do a release of 1.3 some time on Wednesday.

File: ccdproc/tests/test_ccdproc.py [hunk 9]
```diff
@@ -481,6 +481,42 @@ def test_flat_correct_min_value(ccd_data):
...
+    # the mean of the flat frame.
+
+    # create the flat, with some scatter
+    # Note that mean value of flat is set below and is different than
+    # the mean of the flat data.
+    flat_mean = 5.0
+    data = np.random.normal(loc=1.0, scale=0.05, size=ccd_data.shape)
+    flat = CCDData(data, meta=fits.header, unit=ccd_data.unit)
+    flat_data = flat_correct(ccd_data, flat, add_keyword=None,
+                             mean_value=flat_mean)
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: should be `norm_value`

File: ccdproc/tests/test_ccdproc.py [hunk 10]
```diff
@@ -481,6 +481,42 @@ def test_flat_correct_min_value(ccd_data):
...
+def test_flat_correct_mean_value(ccd_data):
+    # Test flat correction with mean value that is different than
+    # the mean of the flat frame.
+
+    # create the flat, with some scatter
+    # Note that mean value of flat is set below and is different than
+    # the mean of the flat data.
+    flat_mean = 5.0
+    data = np.random.normal(loc=1.0, scale=0.05, size=ccd_data.shape)
+    flat = CCDData(data, meta=fits.header, unit=ccd_data.unit)
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: sorry, it should be `fits.Header()`, not `fits.header`.

File: ccdproc/tests/test_ccdproc.py [hunk 11]
```diff
@@ -481,6 +481,42 @@ def test_flat_correct_min_value(ccd_data):
...
+def test_flat_correct_mean_value_bad_value(ccd_data):
+    # Test that flat_correct raises the appropriate error if
+    # it is given a bad mean_value. Bad means <=0.
+
+    # create the flat, with some scatter
+    data = np.random.normal(loc=1.0, scale=0.05, size=ccd_data.shape)
+    flat = CCDData(data, meta=fits.header.Header(), unit=ccd_data.unit)
+    with pytest.raises(ValueError) as e:
+        flat_correct(ccd_data, flat, add_keyword=None,
+                     mean_value=-7)
```
Comment (MSeifert04) [inline ccdproc/tests/test_ccdproc.py]: Should be `norm_value` now

**Comment by mwcraig** [comment]
@MSeifert04 -- just a heads up that it will be another 4-5 hours before I can update this...feel free to push changes before then if you want.

**Comment by MSeifert04** [comment]
@mwcraig The tests pass again and I think I replaced all remaining mean_value references with norm_value. I also added this to the 1.3 milestone and will merge it later if you or @crawfordsm don't object. :)

**Comment by MSeifert04** [comment]
Thanks @bjweiner and @mwcraig!
